Add option for showing the overlay behind the menu

realnc opened the pull request:
## Description

Unlike input overlays, for purely decorative overlays like bezels and borders, it doesn't make sense to have the overlay obscure the menu. This option makes it possible to have the overlay shown behind the menu.

The new option ("Show Overlay Behind Menu") in the "On-Screen Overlay" settings menu. The default of OFF, which is the current behavior:

![2021-12-15-23-19-42-917672197](https://user-images.githubusercontent.com/1665903/146266663-b7c8ea7c-b4a4-4127-8478-e78dae02f6cd.jpg)

When setting it on ON, you get this instead:

![2021-12-15-23-19-52-584955892](https://user-images.githubusercontent.com/1665903/146266684-dc46d881-0eb3-4140-bb53-d83d6e8bbb3a.jpg)

The commits are split in a way that allows for reverting support in individual drivers. The first commit adds the infrastructure for this option without implementing it in any driver. The later commits then implement this option in the gl1, gl2, gl3, vulkan, d3d9, d3d10, d3d11, d3d12 and ctr (3ds) drivers.

For the rest of the drivers, other people will need to implement it. I can't test any of them.

## Reviewers

@jdgleaver @twinaphex @m4xw @RobLoach

MrHuu commented:
This would also be a good addition to the 3ds platform.
As it would keep the overlay loaded when entering the menu, instead of having to reload the overlay when returning from menu.
Currently the option to hide the overlay in menu unloads the overlay and reloads when needed.
With the limited i/o speeds reading from sd, loading usually isn't instant.
Seems like a good option to have available. Next to improved visuals when using menu transparency.

jdgleaver commented:
Oooh - Very nice option!

> I'm not sure that querying the option with config_get_ptr() in the video drivers is the correct way to do this

That's essentially the only issue here. It is bad practice to access the settings pointer in any of the `_frame()` functions - instead, a new `overlay_above_menu` field should be added to the `video_frame_info_t` struct, which you can populate in this function: https://github.com/libretro/RetroArch/blob/1582856b43656ad277b267cc9c36ad6371e8a4bd/gfx/video_driver.c#L2672

> Also, this new option seems to make the existing "hide overlay in menu" option redundant. If you can choose to show the overlay behind the menu, then it seems to me there would be no reason to disable the overlay while in the menu anymore?

I agree with you. "hide overlay in menu" should almost certainly be removed once this PR is ready to go.

realnc commented:
> This would also be a good addition to the 3ds platform.

Are you able to build RA for 3ds and test?

realnc commented:
> That's essentially the only issue here. It is bad practice to access the settings pointer in any of the `_frame()` functions - instead, a new `overlay_above_menu` field should be added to the `video_frame_info_t` struct

Yeah, that's cleaner then. I've done that now.

> > Also, this new option seems to make the existing "hide overlay in menu" option redundant.
>
> I agree with you. "hide overlay in menu" should almost certainly be removed once this PR is ready to go.

The issue is that not all drivers support it. How do you recommend handling this? Check if the currently active video driver is supported and only then expose the new option in the UI (and hide the other one)?

jdgleaver commented:
> The issue is that not all drivers support it. How do you recommend handling this? Check if the currently active video driver is supported and only then expose the new option in the UI (and hide the other one)?

Yes - if it can't be enabled for all gfx drivers, then the approach you suggest is 100% correct.

realnc commented:
OK, so I added `has_overlay_behind_menu` as bool member of `video_driver_t` where drivers can report whether or not they support this. This is set to true on gl1/gl2/gl3/vulkan, all others set it to false. This can be queried in `menu_setting.c` where the option is omitted from the menu for drivers that don't have support.

Is this the correct approach?

jdgleaver commented:
> OK, so I added has_overlay_behind_menu as bool member of video_driver_t where drivers can report whether or not they support this. This is set to true on gl1/gl2/gl3/vulkan, all others set it to false. This can be queried in menu_setting.c where the option is omitted from the menu for drivers that don't have support.
> Is this the correct approach?

Ah, I'm afraid I'm going to have to annoy you with more RA foibles...

For gfx driver supported features, we normally use the `get_flags` function - e.g.:

https://github.com/libretro/RetroArch/blob/cd50a51f3e71101311a62430d44a12989dc79249/gfx/drivers/gl3.c#L2030-L2040

https://github.com/libretro/RetroArch/blob/cd50a51f3e71101311a62430d44a12989dc79249/gfx/video_defines.h#L195-L210

https://github.com/libretro/RetroArch/blob/cd50a51f3e71101311a62430d44a12989dc79249/menu/menu_setting.c#L12330-L12346

So you could just add a new `GFX_CTX_FLAGS_OVERLAY_BEHIND_MENU` enum there

MrHuu commented:
> > This would also be a good addition to the 3ds platform.
> 
> Are you able to build RA for 3ds and test?

I like to think i can.

I've created a PR with the required changes: https://github.com/realnc/RetroArch/pull/1

realnc commented:
OK, I've changed the way support is indicated. This is now done through a new `GFX_CTX_FLAGS_OVERLAY_BEHIND_MENU_SUPPORTED` flag of the `display_flags` enum.

I kept @MrHuu 's commit that adds support to the 3DS driver separate, so it can easily be reverted if it turns out it doesn't work.

The automatic hiding of the "hide overlay in menu" option turns out to be a can of worms. So I claim it's out of scope for this PR :stuck_out_tongue: 

I removed the draft status of the PR. So it's ready, unless there's any other issues.

MrHuu commented:
Latest changes also work like a charm on 3DS hardware. 
Much appreciated functionality, thanks.

MrHuu commented on CHANGES.md:
```diff
@@ -1,5 +1,6 @@
 # Future
 - LINUX: Added support for Linux GameMode (https://github.com/FeralInteractive/gamemode), which can be toggled on/off in the Power Management or Latency settings menus.
+- Added support for showing the overlay behind the menu instead of in front. This is currently only supported on the GL and Vulkan drivers.
```
```suggestion
- Added support for showing the overlay behind the menu instead of in front. This is currently only supported on the GL, Vulkan and 3DS drivers.
 ```

realnc commented:
> Latest changes also work like a charm on 3DS hardware. Much appreciated functionality, thanks.

Thanks for testing!

jdgleaver commented:
Excellent - this looks perfect now.

> The automatic hiding of the "hide overlay in menu" option turns out to be a can of worms. So I claim it's out of scope for this PR stuck_out_tongue (Just hiding the option is not enough. It would also need to be ignored if already set, which results in a lot of if throughout the code. Maybe there's an elegant way to do that, but I can't find it.)

I agree 100%. This needs some careful consideration, and it is most certainly out of scope for the current PR.

realnc commented:
@sonninnos Added support for d3d 9/10/11/12. I changed this PR to have individual commits for each driver, while the first commit implements the infrastructure for this option. This should make it easy to revert a specific driver implementation in case it turns out it breaks something without having to nuke the whole PR.
